packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -28,26 +26,120 @@ const protocolVersions = ["^0.4.0", "^0.3.0", "^0.2.0", "^0.1.0"];
+                eventName: "OdspDocumentDeltaCollection.GetSocketIoReference",
+                delayDeleteDelta: socketReference.delayDeleteTimeoutSetTime !== undefined ?
+                    (Date.now() - socketReference.delayDeleteTimeoutSetTime) : undefined,
+            });
+        }
+
+        return socketReference;
+    }
+
+    /**
+     * Removes a reference for the given key
+     * Once the ref count hits 0, the socket is disconnected and removed
+     * @param key - socket reference key
+     * @param isFatalError - true if the socket reference should be removed immediately due to a fatal error
+     */
+    public removeSocketIoReference(isFatalError: boolean) {
+        assert(this.references > 0);
+        this.references--;
+
+        // see comment in disconnected() getter
+        this.inUse = true;
+
+        if (isFatalError || this.disconnected) {
+            this.closeSocket();
+            return;
+        }
+
+        if (this.references === 0 && this.delayDeleteTimeout === undefined) {
+            this.delayDeleteTimeout = setTimeout(() => {
+                // We should not get here with active users.
+                assert(this.references === 0);
+                this.closeSocket();
+            }, socketReferenceBufferTime);
+            this.delayDeleteTimeoutSetTime = Date.now();
+        }
+    }
+
+    public get socket() {
+        if (!this._socket) {
+            throw new Error(`Invalid socket for key "${this.key}`);
+        }
+        return this._socket;
+    }
+
+    public constructor(public readonly key: string, socket: SocketIOClient.Socket) {
+        this._socket = socket;
+        SocketReference.socketIoSockets.set(key, this);
     }
 
-    public clearTimer() {
+    private clearTimer() {
         if (this.delayDeleteTimeout !== undefined) {
             clearTimeout(this.delayDeleteTimeout);
             this.delayDeleteTimeout = undefined;
             this.delayDeleteTimeoutSetTime = undefined;
         }
     }
 
-    public closeSocket() {
-        if (this.socket) {
-            this.socket.disconnect();
-            this.socket = undefined;
-        }
+    private closeSocket() {
+        if (!this._socket) { return; }
+
+        this.clearTimer();
+
+        assert(SocketReference.socketIoSockets.get(this.key) === this);
+        SocketReference.socketIoSockets.delete(this.key);
+
+        const socket = this._socket;
+        this._socket = undefined;
+
+        // Delay closing socket, to make sure all users of socket observe the same event that causes
+        // this instance to close, and thus properly record reason for clusure.
+        // All event raising is synchronous, so clients will have a chance to react before socket is
+        // closed without any extra data on why it was closed.
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        Promise.resolve().then(() => { socket.disconnect(); });
+    }
+
```

Comment:
When the host immediately connects and disconnects, we see that inUse is going to be true because of removeSocketIoReference being called.
Is it possible that we still want to reuse this connection in that scenario?

Reply:
Yes, inUse will become true. And yes, we will continue to reuse socket as long as it's connected. The moment it becomes disconnected, we will not reuse it.
It's easier to think about it from how code differs from old code. The only difference if we socket is disconnected, but inUse = false. In such case, we will keep reusing socket with these changes. Because clients not removing their references means socket never connected in the first place (if it did, then we would see clients removing their references on disconnect)